### ISSUE TITLE
echo doesn't work in namespaces (and other UNIVERSE handling oddities)

### ISSUE THREAD

**Issue by stephane-chazelas** [ksh93/ksh#727]
```
$ ksh --version
  version         sh (AT&T Research) 93u+m/1.1.0-alpha+9a48c61e 2024-03-07
$ ksh -c 'namespace foo { echo foo; }'
ksh: echo: _AST_FEATURES=UNIVERSE - ucb: is not an identifier
```

Work around seems to be to set the `_AST_FEATURES` variable or call  (builtin) `getconf UNIVERSE` which seems to set (and export!) `_AST_FEATURES`, in the namespace or not:

```
$ namespace foo { echo foo; }
ksh: echo: _AST_FEATURES=UNIVERSE - ucb: is not an identifier
$ .foo._AST_FEATURES=
$ namespace foo { echo foo; }
```

```
$ namespace foo { echo foo; }
ksh: echo: _AST_FEATURES=UNIVERSE - ucb: is not an identifier
$ typeset -p _AST_FEATURES
$ command /opt/ast/bin/getconf UNIVERSE
ucb
$ typeset -p _AST_FEATURES
typeset -x _AST_FEATURES='UNIVERSE - ucb'
$ namespace foo { echo foo; }
foo
```

See also:

```
$ echo '\\'
\\
$ namespace ucb { command /opt/ast/bin/getconf UNIVERSE = ucb; }
getconf: _AST_FEATURES=UNIVERSE - ucb: is not an identifier
$ namespace ucb { _AST_FEATURES=; command /opt/ast/bin/getconf UNIVERSE = ucb; }
$ echo '\\'
\\
$ namespace xpg4 { _AST_FEATURES=; command /opt/ast/bin/getconf UNIVERSE = xpg4; }
$ echo '\\'
\
$ typeset -p _AST_FEATURES
typeset -x _AST_FEATURES='UNIVERSE - ucb'
$ namespace xpg4 { typeset -p _AST_FEATURES; }
typeset -x _AST_FEATURES='UNIVERSE = xpg4'
$ namespace ucb { typeset -p _AST_FEATURES; }
typeset -x _AST_FEATURES='UNIVERSE = ucb'
$ command /opt/ast/bin/getconf UNIVERSE
xpg4
$ namespace ucb { command /opt/ast/bin/getconf UNIVERSE; }
xpg4
$ namespace ucb { echo '\\'; }
\
$ namespace xpg4 { echo '\\'; }
\
$ typeset -p {,.ucb.,.xpg4.}_AST_FEATURES
typeset -x _AST_FEATURES='UNIVERSE - ucb'
typeset -x .ucb._AST_FEATURES='UNIVERSE = ucb'
typeset -x .xpg4._AST_FEATURES='UNIVERSE = xpg4'
```

The *UNIVERSE* handling seems not to be handling namespaces properly.

**Comment by McDutchie** [comment]
Yikes. Thanks for the report. This is broken as of commit 0ea4e5599609c976d85f7d28063bdc43ae3ec92b.

The reason I didn't noticed the bug is that simply running "echo" sets (and exports) _AST_FEATURES. This is only checked for the very first time echo is run in a ksh process. So the bug is only reproducible if `echo` is run in a namespace before having been run anywhere else.

Reliable reproducer:

```
$ (unset _AST_FEATURES; ksh -c 'namespace foo { echo test; }')
ksh: echo: _AST_FEATURES=UNIVERSE - ucb: is not an identifier
```

For 1.1+ I'm planning to remove the `UNIVERSE` handling from the `echo` command; in my view, the built-in should act identically regardless of the OS. But this should still be fixed for 1.0.

**Comment by McDutchie** [comment]
The specific code change in 0ea4e5599609c976d85f7d28063bdc43ae3ec92b that triggers the bug is the change from `dtinsert` to `dtinstall` in nv_search(). Change that back to `dtinsert` and the reproducer works again.

src/lib/libast/man/cdt.3:
```
   OBJECT OPERATIONS
     void* dtinsert(Dt_t* dt, void* obj)
     void* dtinstall(Dt_t* dt, void* obj)
     void* dtappend(Dt_t* dt, void* obj)
       These functions add an object prototyped by obj into dt.  See
       Dtdisc_t.makef for object construction.  dtinsert() and dtappend()
       perform the same function for all methods except for Dtlist (see Dtlist
       for details).  For Dtset, Dtrhset or Dtoset, if there is an object in dt
       matching obj, dtinsert() and dtappend() will not insert a new object,
       whereas dtinstall() will remove such a matching object before inserting
       the new object.
       
       On failure, dtinsert() and dtinstall() return NULL.  Otherwise, the
       return value is either the newly inserted object or the matching object
       as noted.
```

I don't know exactly why this change to dtinstall triggers the bug, but I don't think there is anything wrong with it; the node was not found, a new one is created, so if there is an old matching node after all (for whatever reason), it *should* be removed or we have a memory leak.

Instead, it seems to have exposed a problem with the shell intercept of getenv(3) and putenv(3): they fail to unset the namespace. libast calls putenv to set the _AST_FEATURES variable.

Patch to test:

```diff
diff --git a/src/cmd/ksh93/sh/name.c b/src/cmd/ksh93/sh/name.c
index 356203159..ad9d1e4a6 100644
--- a/src/cmd/ksh93/sh/name.c
+++ b/src/cmd/ksh93/sh/name.c
@@ -2998,12 +2998,19 @@ static char *oldgetenv(const char *string)
  */
 char *sh_getenv(const char *name)
 {
-	Namval_t *np;
+	Namval_t *np, *savns;
+	char *cp, *savpr;
 	if(!sh.var_tree)
 		return oldgetenv(name);
-	else if((np = nv_search(name,sh.var_tree,0)) && nv_isattr(np,NV_EXPORT))
-		return nv_getval(np);
-	return NULL;
+	/* deactivate a possible namespace or compound assignment */
+	savns = sh.namespace, savpr = sh.prefix;
+	sh.namespace = NULL, sh.prefix = NULL;
+	if((np = nv_search(name,sh.var_tree,0)) && nv_isattr(np,NV_EXPORT))
+		cp = nv_getval(np);
+	else
+		cp = NULL;
+	sh.namespace = savns, sh.prefix = savpr;
+	return cp;
 }
 
 #ifndef _NEXT_SOURCE
@@ -3027,9 +3034,14 @@ int putenv(const char *name)
 	Namval_t *np;
 	if(name)
 	{
+		Namval_t *savns = sh.namespace;
+		char *savpr = sh.prefix;
+		/* deactivate a possible namespace or compound assignment */
+		sh.namespace = NULL, sh.prefix = NULL;
 		np = nv_open(name,sh.var_tree,NV_EXPORT|NV_IDENT|NV_NOARRAY|NV_ASSIGN);
 		if(!strchr(name,'='))
 			_nv_unset(np,0);
+		sh.namespace = savns, sh.prefix = savpr;
 	}
 	return 0;
 }
@@ -3042,11 +3054,12 @@ char* sh_setenviron(const char *name)
 	Namval_t *np;
 	if(name)
 	{
-		char *save_prefix = sh.prefix;
-		/* deactivate a possible compound assignment */
-		sh.prefix = NULL;
+		Namval_t *savns = sh.namespace;
+		char *savpr = sh.prefix;
+		/* deactivate a possible namespace or compound assignment */
+		sh.namespace = NULL, sh.prefix = NULL;
 		np = nv_open(name,sh.var_tree,NV_EXPORT|NV_IDENT|NV_NOARRAY|NV_ASSIGN);
-		sh.prefix = save_prefix;
+		sh.namespace = savns, sh.prefix = savpr;
 		if(strchr(name,'='))
 			return nv_getval(np);
 		_nv_unset(np,0);
```

**Comment by McDutchie** [comment]
See also dfe836d44c775a49d31c17e4f30680b428ef98f7.

**Comment by stephane-chazelas** [comment]
2024-03-10 05:39:33 -0700, Martijn Dekker:
[...]
> For 1.1+ I'm planning to remove the `UNIVERSE` handling from
> the `echo` command; in my view, the built-in should act
> identically regardless of the OS. But this should still be
> fixed for 1.0.
[...]

I don't know what else the UNIVERSE affects, but having a way to
customise the behaviour of echo like with the bsdecho option of
zsh or xpg_echo option of bash or $ECHO_STYLE variable of yash
could still be useful.
